### ISSUE TITLE
File checksum/file size does not update after anonymization

### ISSUE THREAD

**Issue by brianzhouzc** [BCCF-UBCO-AD/Orthanc-TMI#88]
As title. Orthanc seems to calculate both the checksum and the file size before it is being handled by our custom storage area.

This means we either have to 1. find a function that allows us to override the default checksum behavior or 2. to just update the checksum and file size in the database ourselves.

There is something call `_OrthancPluginComputeHash` in the plugin header but I can't find any documentation on it, doubt it's useful. Updating the checksum ourselves is probably eaiser.

**Comment by brianzhouzc** [comment]
I discover that the checksum must've been done elsewhere after `WriteDicomFile`. The uuid of the fil doesn't exist in the database during `WriteDicomFile`. I don't know why this surprises me lol, it's so obvious now looking back.

The problem now is, where does the checksum or the file data get written to the database? The checksum for the file written to the database is still the original md5 and size, before the filtering is done.

**Comment by cppcooper** [comment]
What's the situation if we use the filter callback instead of storage area callbacks (or in addition to)

**Comment by brianzhouzc** [comment]
> What's the situation if we use the filter callback instead of storage area callbacks (or in addition to)

I'll investigate that

**Comment by brianzhouzc** [comment]
`OrthancPluginIncomingDicomInstanceFilter` receives a pointer to an `OrthancPluginDicomInstance` object, which means that everything should be in the database when it is called. I don't see a way to get the UUID of the file for the DicomInstance tho..

**Comment by brianzhouzc** [comment]
I take it back. No nothing should be in the database yet, but this should be the function that we do the anonymization on. Since it receives a pointer to DicomInstance we can just modify the instance directly. What do you think? @cppcooper

**Comment by cppcooper** [comment]
I think there is a commit somewhere in the history that does this. It sounds good, I'm just not sure it works like that cause I haven't looked at it since early November I think

**Comment by brianzhouzc** [comment]
> I think there is a commit somewhere in the history that does this. It sounds good, I'm just not sure it works like that cause I haven't looked at it since early November I think

I saw there's a commented line with `OrthancPluginIncomingDicomInstanceFilter` in the current develop branch, I will look through the commits as well.

The one concern I have is there isn't a way to modify the size of said instance. Maybe instead of modifying the instance we discard it and replace it with our new instance.

**Comment by brianzhouzc** [comment]
There's also this call back that we can maybe use: [OrthancPluginOnStoredInstanceCallback](https://sdk.orthanc-server.com/group__Callbacks.html#gaa1135f25510f01372d746b54dafd3f6e)

**Comment by cppcooper** [comment]
```
orthanc-server  | W0128 10:09:12.422416 PluginsManager.cpp:168] OnIncomingCallback: 139944353807200
...
orthanc-server  | W0128 10:09:12.426424 PluginsManager.cpp:168] WriteDicomFile: success
orthanc-server  | W0128 10:09:12.434887 PluginsManager.cpp:168] OnStoredCallback: 139944353807184
```
So we can see here the stored callback happens after the file is saved. This is probably when the db is updated.

So we could try to update the DB in the `OnStoredCallback`, it will hopefully have the data.. if not then we'll have to try this other option.

Or we could move the `ApplyFilter` call to `OnIncomingCallback` and try to create a new Dicom Instance and discard the incoming version.. then simplify the storage area to just write the file and perform hard linking.

**Comment by cppcooper** [comment]
So based on this morning's tests option 2 will require more than
```cpp
    auto i2 = OrthancPluginCreateDicomInstance(globals::context, buffer.get(), size);
    OrthancPluginFreeDicomInstance(globals::context,i2);
```
The core managed instance isn't actually stored, so we might need a library to perform a C-store operation to the localhost server. Unless we can find an alternative in the Orthanc api. So let's see about doing the update in the on stored callback first so the data maintains the one way trip aspect we've got now.

**Comment by brianzhouzc** [comment]
Sounds like a plan, I'm working on a POC right now, should be up in the repo tonight.

**Comment by brianzhouzc** [comment]
> ```
> orthanc-server  | W0128 10:09:12.422416 PluginsManager.cpp:168] OnIncomingCallback: 139944353807200
> ...
> orthanc-server  | W0128 10:09:12.426424 PluginsManager.cpp:168] WriteDicomFile: success
> orthanc-server  | W0128 10:09:12.434887 PluginsManager.cpp:168] OnStoredCallback: 139944353807184
> ```
> 
> So we can see here the stored callback happens after the file is saved. This is probably when the db is updated.
> 
> So we could try to update the DB in the `OnStoredCallback`, it will hopefully have the data.. if not then we'll have to try this other option.
> 
> Or we could move the `ApplyFilter` call to `OnIncomingCallback` and try to create a new Dicom Instance and discard the incoming version.. then simplify the storage area to just write the file and perform hard linking.

How did you get this result? I'm using `OrthancPluginGetInstanceSize` to retrieve the size. However, the size is still wrong (i.e same size as the unfiltered file). Using `OrthancPluginComputeMd5` also gives the old, unfiltered md5 as well.

Debug output:
```bash
orthanc-server  | W0131 22:23:29.386636 PluginsManager.cpp:168] OnStored Instance Id: 87ae9004-f692ccad-c4599afc-64b4d6ba-1dff969c
orthanc-server  | W0131 22:23:29.387522 PluginsManager.cpp:168] Checksum: MD5 = ee4dbf69191ef6e3c2c088083a4e2f8e, size = 577536
```

Database:
```bash
uncompressedsize | uncompressedhash
577536 | ee4dbf69191ef6e3c2c088083a4e2f8e
```

**Comment by cppcooper** [comment]
> Sounds like a plan, I'm working on a POC right now, should be up in the repo tonight.

Point of contact? Proof of concept?

> How did you get this result? 
What do you mean? What result? All I did was register callbacks and add logging to see what order they happen in. The numbers are pointers.

> I'm using `OrthancPluginGetInstanceSize` to retrieve the size. However, the size is still wrong (i.e same size as the unfiltered file). Using `OrthancPluginComputeMd5` also gives the old, unfiltered md5 as well.
Not sure if it was in this group, but I've got an md5 function we can use and just compute it on the buffer directly. It would make sense that Orthanc would only calculate the md5 once and return the cached result any time something asks for it.

**Comment by cppcooper** [comment]
I suppose this can be closed with #99 and #113
I think a bug with the checksum was fixed in the latter.

**Comment by brianzhouzc** [comment]
This is fixed
